### DISCUSSION TITLE
IMU Kalibr parameters for AirSim

### DISCUSSION THREAD

**Discussion by alberto1995** [microsoft/AirSim#3523] [Support Q&A]
### What's your question?
Hi everyone,
I'm trying to run some VIO algorithms (Open-VINS and VINS-Mono) using a rosbag recorded with the drone of AirSim. I'm here to ask if someone knows, and can share, IMU noise parameters in [Kalibr](https://github.com/ethz-asl/kalibr/wiki/IMU-Noise-Model) format.

### Include context on what you are trying to achieve

I would like to run VIO algorithms with the exact IMU noise parameters needed by Open-VINS and VINS-Mono (Gyroscope "white noise" - Accelerometer "white noise" - Gyroscope "random walk" - Accelerometer "random walk").

### Include details of what you already did to find answers
First of all, I have seen that In the [SimpleIMU](https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/imu/ImuSimpleParams.hpp#L18) code there are four parameters setted to emulate the MPU-6000, but them don't match with the Kalibr format. Also a Datasheet of the IMU is given, but again I can't find IMU noise parameters in Kalibr format. Secondly, I have seen that other issues #2369 #3269 #2836 have been opened and are still pending, so the pubblication of those parameters can help a lot.

Thank you for your support,
sincerely,

Alberto

**Comment by karnikram** [comment]
Hi @alberto1995, I had the same issue.

I was able to infer the noise model parameters from these two scripts - [ImuSimpleParams](https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/imu/ImuSimpleParams.hpp), [ImuSimple](https://github.com/microsoft/AirSim/blob/238c8e708dab831bf5835b1c954abfb0136e9277/AirLib/include/sensors/imu/ImuSimple.hpp#L71)

`gyro.arw` is the `gyroscope_noise_density`, and is equal to `8.7266462e-5`
`gyro_bias_stability_norm` is the `gyroscope_random_walk`, and is equal to `9.9735023e-7`
`accel.vrw` is the `accelerometer_noise_density`, and is equal to `0.002353596`
`accel_bias_stability_norm` is the `accelerometer_random_walk`, and is equal to `1.2481827e-5`

I also verified these values using [kalibr_allan](https://github.com/rpng/kalibr_allan) after recording a four hour static sequence, and the values are very similar:

![gyro](https://user-images.githubusercontent.com/12653355/106905614-76866400-6722-11eb-9438-f03d669f79e2.png)
![accel](https://user-images.githubusercontent.com/12653355/106905617-771efa80-6722-11eb-8dfe-530677413079.png)


I'm not sure about Open-VINS, but VINS-Mono expects these parameters in discrete-time form. Meaning, the noise density parameters need to be divided by the square root of your sampling time, and random walk parameters need to be multiplied with the square root of your sampling time. Note that the airsim sampling time has an [upper limit](https://github.com/microsoft/AirSim/blob/238c8e708dab831bf5835b1c954abfb0136e9277/AirLib/include/sensors/imu/ImuSimple.hpp#L78) of 0.001 secs. More over in my case I also noticed a time offset between camera and IMU which I had to separately calibrate using Kalibr.

Hope this helps

**Reply by catproof** [reply]
how would you get the transformation from camera to body frame? I changed the position of my front camera, by moving it along the X axis. (moving the camera forward)

**Reply by catproof** [reply]
you can see my post here:
https://github.com/microsoft/AirSim/issues/4119

**Reply by karnikram** [reply]
The position and orientation of the camera is specified in the settings.json file with respect to the IMU frame.

For your case it would just be [1, 0, 0, 0.35; 0, 1, 0, 0; 0, 0, 1, 0; 0, 0, 0, 1]. If you want the transformation of IMU with respect to camera, it's just the inverse of this matrix.

**Reply by karnikram** [reply]
In our case though in addition to the spatial transform we also noticed a slight time offset between the camera and IMU while recording in AirSim. We later estimated that by simulating a checkerboard sequence and using Kalibr.

**Reply by catproof** [reply]
I see. according to [this user](https://github.com/microsoft/AirSim/issues/2786#issuecomment-953122057), the transformation matrix should have some rotations, because the IMU and camera use different coordinate systems. as in, the rotation matrix would be
[0,0,1,
 1,0,0,
 0,1,0]

**Comment by GimpelZhang** [comment]
The two VIO, open_vins and vins_mono, expect different forms of IMU parameters. 

* open_vins: continuous time noise parameters

[evidence:](https://github.com/rpng/open_vins/blob/master/ov_msckf/src/state/Propagator.cpp#L381-L392)

```
    // Construct our discrete noise covariance matrix
    // Note that we need to convert our continuous time noises to discrete
    // Equations (129) amd (130) of Trawny tech report
    Eigen::Matrix<double,12,12> Qc = Eigen::Matrix<double,12,12>::Zero();
    Qc.block(0,0,3,3) = _noises.sigma_w_2/dt*Eigen::Matrix<double,3,3>::Identity();
    Qc.block(3,3,3,3) = _noises.sigma_a_2/dt*Eigen::Matrix<double,3,3>::Identity();
    Qc.block(6,6,3,3) = _noises.sigma_wb_2*dt*Eigen::Matrix<double,3,3>::Identity();
    Qc.block(9,9,3,3) = _noises.sigma_ab_2*dt*Eigen::Matrix<double,3,3>::Identity();

    // Compute the noise injected into the state over the interval
    Qd = G*Qc*G.transpose();
    Qd = 0.5*(Qd+Qd.transpose());
```

where sigma_w_2/sigma_a_2/sigma_wb_2/sigma_ab_2 are the squares of the IMU param inputs: [from ros](https://github.com/rpng/open_vins/blob/master/ov_msckf/src/utils/parse_ros.h#L153-L157)

* vins_mono: as @karnikram pointed out, discrete time noise parameters

evidence:
[construct noise matrix](https://github.com/HKUST-Aerial-Robotics/VINS-Mono/blob/master/vins_estimator/src/factor/integration_base.h#L21-L27):
```
    noise = Eigen::Matrix<double, 18, 18>::Zero();
    noise.block<3, 3>(0, 0) =  (ACC_N * ACC_N) * Eigen::Matrix3d::Identity();
    noise.block<3, 3>(3, 3) =  (GYR_N * GYR_N) * Eigen::Matrix3d::Identity();
    noise.block<3, 3>(6, 6) =  (ACC_N * ACC_N) * Eigen::Matrix3d::Identity();
    noise.block<3, 3>(9, 9) =  (GYR_N * GYR_N) * Eigen::Matrix3d::Identity();
    noise.block<3, 3>(12, 12) =  (ACC_W * ACC_W) * Eigen::Matrix3d::Identity();
    noise.block<3, 3>(15, 15) =  (GYR_W * GYR_W) * Eigen::Matrix3d::Identity();
```

[use the noise in the covariance prediction](https://github.com/HKUST-Aerial-Robotics/VINS-Mono/blob/master/vins_estimator/src/factor/integration_base.h#L125):
```
    covariance = F * covariance * F.transpose() + V * noise * V.transpose();
```
where ACC_N/GYR_N/ACC_W/GYR_W are the IMU param inputs: [read settings](https://github.com/HKUST-Aerial-Robotics/VINS-Mono/blob/master/vins_estimator/src/parameters.cpp#L70-L73)

We can see in vins_mono, there is no division or multiplication of **dt**.

**Reply by catproof** [reply]
@karnikram I am not completely convinced you listed the right parameters. 

If you look at https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/imu/ImuSimpleParams.hpp#L54 you can see that this is angular random walk. This is taken straight from Kalibr Allan: "The gyroscope_random_walk and accelerometer_random_walk values can normally be found on the IMU datasheet as either angular random walk or velocity random walk, respectively."
That means that gyro.arw should actually be gyroscope_random_walk. The same goes for velocity random walk... accel.vrw is accelerometer_random_walk.

I am not sure what the noise parameters are though.

**Reply by catproof** [reply]
Also, how did you set up Kalibr Allan in AirSim? don't you need to point a camera at an April grid?....

**Reply by catproof** [reply]
@karnikram  ok so on the Kalibr Github, it says something that makes sense with what you said, it says:
"The parameters for the "white noise" processes (, ) are often specified in the datasheet of the sensor manufacturer. A bit misleading, they are commonly denoted as angular random walk in case of the gyro, and velocity random walk for the accel."

also, I am now realizing you got the values using a static sequence (in which case, you mean you just recorded a drone staying still, to get the noise parameters instead of using an April grid).

**Reply by karnikram** [reply]
Yes, that is correct. I kept the drone static and just recorded the IMU values for Kalibr Allan.
